predicators/refinement_estimators/per_skeleton_estimator.py

```diff
@@ -0,0 +1,66 @@
+"""A learning-based refinement cost estimator that trains and stores a separate
+cost predictor for each unique (skeleton, atoms_sequence) key."""
+
+import abc
+from pathlib import Path
+from typing import Dict, FrozenSet, Generic, List, Optional, Set, Tuple, \
+    TypeVar
+
+import dill as pkl
+
+from predicators.refinement_estimators import BaseRefinementEstimator
+from predicators.structs import GroundAtom, Task, _GroundNSRT
+
+# Type of the (skeleton, atoms_sequence) key for model dictionary
+# which converts both of them to be immutable
+ModelDictKey = Tuple[Tuple[_GroundNSRT, ...],  # skeleton converted to tuple
+                     Tuple[FrozenSet[GroundAtom], ...]  # atoms_sequence
+                     ]
+T = TypeVar('T')
+
+
+class PerSkeletonRefinementEstimator(BaseRefinementEstimator, Generic[T]):
+    """A refinement cost estimator that trains a separate cost predictor per
+    skeleton, which is given as a (skeleton, atoms_sequence) pair."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        # _model_dict maps immutable skeleton atoms_sequence pair to model
+        self._model_dict: Optional[Dict[ModelDictKey, T]] = None
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    def get_cost(self, initial_task: Task, skeleton: List[_GroundNSRT],
+                 atoms_sequence: List[Set[GroundAtom]]) -> float:
+        assert self._model_dict is not None, "Need to train"
+        key = self._immutable_model_dict_key(skeleton, atoms_sequence)
+        # If key isn't in dictionary (no data for skeleton), cost is infinite
+        if key not in self._model_dict:
+            return float('inf')
+        model = self._model_dict[key]
+        cost = self._model_predict(model, initial_task)
+        return cost
+
+    @abc.abstractmethod
+    def _model_predict(self, model: T, initial_task: Task) -> float:
+        """Get the cost prediction from a model using the initial task as
+        input."""
```

Comment:
wait we don't need to `raise NotImplementedError`? lol why are we doing this everywhere else in the code... maybe i'll change it someday